engine/api/application_import.go

```diff
@@ -0,0 +1,163 @@
+
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			return sdk.NewError(sdk.ErrWrongRequest, err)
+		}
+		defer r.Body.Close()
+
+		contentType := r.Header.Get("Content-Type")
+		if contentType == "" {
+			contentType = http.DetectContentType(body)
+		}
+
+		var eapp = new(exportentities.Application)
+		var errapp error
+		switch contentType {
+		case "application/json":
+			errapp = json.Unmarshal(body, eapp)
+		case "application/x-yaml", "text/x-yam":
+			errapp = yaml.Unmarshal(body, eapp)
+		default:
+			return sdk.NewError(sdk.ErrWrongRequest, fmt.Errorf("unsupported content-type: %s", contentType))
+		}
+
+		if errapp != nil {
+			return sdk.NewError(sdk.ErrWrongRequest, err)
+		}
+
+		log.Info("postApplicationImportHandler> Import application %s in project %s (force=%v)", eapp.Name, key, force)
+		log.Debug("postApplicationImportHandler> App: %+v", eapp)
+
+		//Check valid application name
+		regexp := regexp.MustCompile(sdk.NamePattern)
+		if !regexp.MatchString(eapp.Name) {
+			return sdk.WrapError(sdk.ErrInvalidApplicationPattern, "postApplicationImportHandler> Application name %s do not respect pattern %s", eapp.Name, sdk.NamePattern)
+		}
+
+		//Load project
+		proj, errp := project.Load(api.mustDB(), api.Cache, key, getUser(ctx), project.LoadOptions.WithGroups)
+		if errp != nil {
+			return sdk.WrapError(errp, "postApplicationImportHandler> Unable load project")
+		}
+
+		//Check if app exist
+		oldApp, errl := application.LoadByName(api.mustDB(), api.Cache, key, eapp.Name, nil, application.LoadOptions.WithVariablesWithClearPassword, application.LoadOptions.WithKeys)
+		if errl != nil && sdk.ErrorIs(errl, sdk.ErrApplicationNotFound) {
+			return sdk.WrapError(errl, "postApplicationImportHandler> Unable to load application")
+		}
+
+		//If the application exist and we don't want to force, raise an error
+		if oldApp != nil && !force {
+			return sdk.ErrApplicationExist
+		}
+
+		//Craft the application
+		app := new(sdk.Application)
+		app.Name = eapp.Name
+		app.VCSServer = eapp.VCSServer
+		app.RepositoryFullname = eapp.RepositoryName
+
+		//Inherit permissions from project
+		if len(eapp.Permissions) == 0 {
+			eapp.Permissions = make(map[string]int)
+			for _, p := range proj.ProjectGroups {
+				eapp.Permissions[p.Group.Name] = p.Permission
+			}
+		}
+
+		//Compute permissions
+		for g, p := range eapp.Permissions {
+			perm := sdk.GroupPermission{Group: sdk.Group{Name: g}, Permission: p}
+			app.ApplicationGroups = append(app.ApplicationGroups, perm)
+		}
+
+		//Compute parameters
+		for p, v := range eapp.Variables {
+			if v.Type == "" {
+				v.Type = sdk.StringVariable
+			}
+			param := sdk.Variable{Name: p, Type: v.Type, Value: v.Value}
+			app.Variable = append(app.Variable, param)
+		}
+
+		tx, errtx := api.mustDB().Begin()
+		if errtx != nil {
+			return sdk.WrapError(errtx, "postApplicationImportHandler> Unable to start transaction")
+
+		}
+
+		defer tx.Rollback()
+
+		done := make(chan bool)
```

Comment:
I have read that a best practice when we use a chan to inform that we're done it's better to use `chan struct{}` less memory space and more explicit that we don't care about the value

Reply:
it's copy & paste from old code, i will change it for a waitgroup